lightning/src/routing/router.rs

```diff
@@ -1512,6 +1516,78 @@ where L::Target: Logger {
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..8]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; 8];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length { 
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
+									cur_node_id = random_channel.node_two;
+									if let Some(dir_info) = &random_channel.one_to_two {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								} else {
+									cur_node_id = random_channel.node_one;
+									if let Some(dir_info) = &random_channel.two_to_one {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								};
+							}
+						}
+					}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40, i.e., the most used cltv_expiry_delta
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the offset to reduce the payment failure probability
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 432; // 3*144
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET);
+
+		// Add 'shadow' CLTV offset to all hops but the final one
+		for h in path {
+			if let Ok(hop) = h {
+				if hop.pubkey != payment_params.payee_pubkey {
+					hop.cltv_expiry_delta = hop.cltv_expiry_delta
+						.checked_add(shadow_ctlv_expiry_delta_offset).unwrap_or(hop.cltv_expiry_delta);
+					let max_total_cltv_expiry_delta = payment_params.max_total_cltv_expiry_delta;
+					if hop.cltv_expiry_delta > max_total_cltv_expiry_delta { 
+						return Err(LightningError{err: "Chosen path exceeds max total CLTV delta limit".to_owned(), action: ErrorAction::IgnoreError});
```

Comment:
One alternative instead of returning an error would be to compute first the sum of the CLTV deltas along the payment path and ensure if added to the sum of the offsets it's not more than the path upper bound. If so, the overhead can be divided by the number of hops and the result subtract on each offset.

This is motivated by the fact that 432, isn't that much. IIRC Eclair nodes are (were?) using 144 as a default CLTV delta value.

Reply:
Yes, this would go hand-in-hand with [Matt's comment](https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r806135532) above, i.e., we would enforce a hard limit and essentially distribute the available 'offset budget' equally over all hops. This could be a feasible way to go forward, however, as already stated there I'm generally a bit critical of enforcing a hard limit, since it may give adversaries additional (potentially even certain) information on the payee. Also, if too 'crooked' values would result from splitting the 'offset budget', an adversary may rightfully be able to reduce the candidate set. So we probably should stick with offsets that are multiples of 40 or 144.

Btw. the `432` value is not thought to be the overall CLTV delta limit, which currently defaults to `1008`, but just an upper bound on the added offset.